general/glossary.md

```diff
@@ -0,0 +1,53 @@
+# Glossary
+
+This documents aims at providing a common lexicon for terms used to refer to artifacts.
+
+### Motivation
+
+Throughout the documentation there are several terms which are often abused because we either have no alternative or there's not enough consensus on what's their canonical name.
+
+This leads to often ambiguous documentation because one term is used for different things or because we use different terms for the same artifact/object.
+
+If this document succeeds, we should end up having a documentation where each object/artifact has one single term to designate it. In other words, [Wittgenstein](https://en.wikipedia.org/wiki/Ludwig_Wittgenstein) should be proud of us (see [Tractatus Logico-Philosophicus](https://en.wikipedia.org/wiki/Tractatus_Logico-Philosophicus)).
+
+Note that we also include an [appendix section](#appendix-trademarks--product-names) listing the canonical spelling and casing for used trademarks or product names.
+
+## Technical Terms
+
+1. Frontend realm:
+    1. **JavaScript module**: a generic JavaScript module in any format (AMD, CommonJS, ...).
+        1. **AMD module**: a JavaScript AMD module (the `.js` file or its logical representation at runtime).
+        2. **CommonJS module**: a JavaScript CommonJS module (the `.js` files used in node).
+        3. **ES6+ module**: EcmaScript 6 or higher code (i.e.: `import from` syntax).
+        4. **Harmony module**: same as ES6+ module (better use _ES6+ module_ instead).
+    2. **npm package**: a JavaScript package (those described by a `package.json` file).
+    3. **Webpack bundle**: a `.js` file created by `webpack`, containing several bundled `.js` files.
+2. Backend realm:
+    1. **Java package**: a Java package as defined by the language.
+    2. **OSGi bundle**: the JAR file deployed to DXP or its logical representation at runtime.
```

Comment:
Maybe we can use them interchangeably. :thinking: 

See https://www.osgi.org/developer/what-is-osgi/, where they define OSGi as "The Dynamic Module System for Java" and https://en.wikipedia.org/wiki/OSGi#Bundles, where they seem to imply that the bundle is the artifact (JAR file with MANIFEST) though I see _bundle_ and _module_ interchanged liberally in all writings related to OSGi.

I think it makes sense to add both as synonyms.

Alternatively, we can use bundle when referring to the JAR file or the Java `Bundle` object and module elsewhere, though I doubt there's any real benefit in this subdivision. Also, it would make writing documentation more difficult.

Reply:
What do you think?

Reply:
I'd go with "bundle" for the JAR (the output) and "module" for the source (eg. `modules/apps/*/*` etc).

- When I'm in the Gogo Shell, I type `lb` to list the bundles (deployed artifacts).
- But when I'm making a change to the source code, I find the project I'm looking for in the `modules` directory.

Another word I often use for these "OSGi" modules is "project", but I am not sure whether that makes sense to anybody else... (eg. "the segments-web project" etc...) Basically trying for something more abstract — "the stuff that Tango is working on" — as opposed to the concrete details of it (am I talking about the fake "npm package" in that diretory with the package.json? or the OSGi module? or am I talking about the JS module pointed at by the `main` field? etc).

🤷

Reply:
Let's split it in two different terms, then :+1: